mm2src/coins/utxo/utxo_common.rs

```diff
@@ -2019,6 +2019,95 @@ pub async fn send_maker_refunds_payment<T: UtxoCommonOps + SwapOps>(
     refund_htlc_payment(coin, args).await.map(|tx| tx.into())
 }
 
+/// Checks if a scriptSig is a script that spends a P2PK output.
+fn does_script_spend_p2pk(script: &Script) -> bool {
+    // P2PK scriptSig is just a single signature. The script should consist of a single push bytes
+    // instruction with the data as the signature.
+    if let Some(Ok(instruction)) = script.get_instruction(0) {
+        match instruction.opcode {
+            Opcode::OP_PUSHBYTES_70 | Opcode::OP_PUSHBYTES_71 | Opcode::OP_PUSHBYTES_72 => {
+                // P2PK scripts just consist of a single signature, there should be no more instructions after the singature.
+                return script.get_instruction(1).is_none();
+            },
+            _ => {},
+        }
+    }
+    false
+}
+
+/// Verifies that the script that spends a P2PK is signed by the expected pubkey.
+fn verify_p2pk_input_pubkey(
+    script: &Script,
+    expected_pubkey: &Public,
+    unsigned_tx: &TransactionInputSigner,
+    index: usize,
+    signature_version: SignatureVersion,
+    fork_id: u32,
+) -> Result<bool, String> {
+    // Extract the signature from the scriptSig.
+    let signature = match script.get_instruction(0) {
+        Some(Ok(instruction)) => match instruction.opcode {
+            Opcode::OP_PUSHBYTES_70 | Opcode::OP_PUSHBYTES_71 | Opcode::OP_PUSHBYTES_72 => match instruction.data {
+                Some(bytes) => {
+                    try_s!(SecpSignature::from_der(&bytes[..bytes.len() - 1]));
+                    bytes.to_vec().into()
+                },
+                None => return ERR!("No data at instruction 0 of script {:?}", script),
+            },
+            _ => return ERR!("Unexpected opcode {:?}", instruction.opcode),
+        },
+        Some(Err(e)) => return ERR!("Error {} on getting instruction 0 of script {:?}", e, script),
+        None => return ERR!("None instruction 0 of script {:?}", script),
+    };
+    // Make sure we have no more instructions. P2PK scriptSigs consist of a single instruction only containing the signature.
+    if script.get_instruction(1).is_some() {
+        return ERR!("Unexpected instruction at position 2 of script {:?}", script);
+    };
+    // Get the scriptPub for this input. We need it to get the transaction sig_hash to sign (but actually "to verify" in this case).
+    let pubkey = expected_pubkey
+        .to_secp256k1_pubkey()
+        .map_err(|e| ERRL!("Error converting plain pubkey to secp256k1 pubkey: {}", e))?;
+    // P2PK scriptPub has two valid possible formats depending on whether the public key is written in compressed or uncompressed form.
+    let possible_pubkey_scripts = vec![
+        Builder::build_p2pk(&Public::Compressed(pubkey.serialize().into())),
+        Builder::build_p2pk(&Public::Normal(pubkey.serialize_uncompressed().into())),
+    ];
+    for pubkey_script in possible_pubkey_scripts {
+        // Get the transaction hash that has been signed in the scriptSig.
```

Comment:
```rust
for pubkey_script in &[
        Builder::build_p2pk(&Public::Compressed(pubkey.serialize().into())),
        Builder::build_p2pk(&Public::Normal(pubkey.serialize_uncompressed().into())),
    ]
```

Reply:
why?

i think that makes it look extra complicated with such multi-line for-loop.

Reply:
you can also just  replace the vec with slice 
let .. = &[
        Builder::build_p2pk(&Public::Compressed(pubkey.serialize().into())),
        Builder::build_p2pk(&Public::Normal(pubkey.serialize_uncompressed().into())),
    ]

Reply:
using vec will allocated on stack while with slices, allocation will be on the stack